File: MaximalSquare5.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    int maximalSquare(vector<vector<char> >& matrix) {
        
        int m = matrix.size();
        if(m == 0) return 0;
        int n = matrix[0].size();
        
        int result = 0;
        int pre = 0;
        vector<int> dp(n, 0);
        
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(j == 0 || i == 0){
                    pre = dp[j];
                    dp[j] = matrix[i][j] - '0';
                }
                else{
                    int t = dp[j];
                    if(matrix[i][j] == '0') dp[j] = 0;
                    else{
                        dp[j] = min(pre, min(dp[j], dp[j-1])) + 1;
                    }
                    pre = t;
                }
                result = max(result, dp[j]);
            }
        }
        
        return result * result;
    }
};

int main(){
    
    int m, n;
    cin>>m>>n;
    
    vector<vector<char> > matrix;
    for(int i=0;i<m;i++){
        vector<char> raw;
        for(int j=0;j<n;j++){
            char ch;
            cin>>ch;
            raw.push_back(ch);
        }
        matrix.push_back(raw);
    }
    
    Solution *solution = new Solution();
    cout<<solution->maximalSquare(matrix);
    
    return 0;
}

File: MaximalSquare5.py
# DP
class Solution:
    def maximalSquare(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype: int
        """
        m = len(matrix)
        if m == 0:
            return 0
        n = len(matrix[0])
        
        result = 0
        dp = [0] * (m + 1)
        pre = 0
        
        for j in range(n):
            for i in range(1, m + 1):
                t = dp[i]
                if matrix[i - 1][j] == '0':
                    dp[i] = 0
                else:
                    dp[i] = min(dp[i], dp[i-1], pre) + 1
                    result = max(result, dp[i])
                pre = t
                
        return result * result
        
if __name__ == '__main__':
    m = int(raw_input())
    matrix = []
    for i in range(m):
        raw = raw_input()
        matrix.append(list(raw))
    solution = Solution()
    print(solution.maximalSquare(matrix))

